src/Neo/SmartContract/Native/Notary.cs

```diff
@@ -0,0 +1,331 @@
+        /// <param name="data">Deposit-related data: optional To value (treated as deposit owner if set) and Till height after which deposit can be withdrawn </param>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.All)]
+        private void OnNEP17Payment(ApplicationEngine engine, UInt160 from, BigInteger amount, StackItem data)
+        {
+            if (engine.CallingScriptHash != GAS.Hash) throw new InvalidOperationException(string.Format("only GAS can be accepted for deposit, got {0}", engine.CallingScriptHash.ToString()));
+            var to = from;
+            var additionalParams = (Array)data;
+            if (additionalParams.Count() != 2) throw new FormatException("`data` parameter should be an array of 2 elements");
+            if (!additionalParams[0].Equals(StackItem.Null)) to = additionalParams[0].GetSpan().ToArray().AsSerializable<UInt160>();
+            var till = (uint)additionalParams[1].GetInteger();
+            var tx = (Transaction)engine.ScriptContainer;
+            var allowedChangeTill = tx.Sender == to;
+            var currentHeight = Ledger.CurrentIndex(engine.SnapshotCache);
+            Deposit deposit = engine.SnapshotCache.GetAndChange(CreateStorageKey(Prefix_Deposit, to))?.GetInteroperable<Deposit>();
+            if (till < currentHeight + 2) throw new ArgumentOutOfRangeException(string.Format("`till` shouldn't be less than the chain's height {0} + 1", currentHeight + 2));
+            if (deposit != null && till < deposit.Till) throw new ArgumentOutOfRangeException(string.Format("`till` shouldn't be less than the previous value {0}", deposit.Till));
+            if (deposit is null)
+            {
+                var feePerKey = Policy.GetAttributeFee(engine.SnapshotCache, (byte)TransactionAttributeType.NotaryAssisted);
+                if ((long)amount < 2 * feePerKey) throw new ArgumentOutOfRangeException(string.Format("first deposit can not be less than {0}, got {1}", 2 * feePerKey, amount));
+                deposit = new Deposit() { Amount = 0, Till = 0 };
+                if (!allowedChangeTill) till = currentHeight + DefaultDepositDeltaTill;
+            }
+            else if (!allowedChangeTill) till = deposit.Till;
+
+            deposit.Amount += amount;
+            deposit.Till = till;
+            PutDepositFor(engine, to, deposit);
+        }
+
+        /// <summary>
+        /// Lock asset until the specified height is unlocked.
+        /// </summary>
+        /// <param name="engine">ApplicationEngine</param>
+        /// <param name="addr">Account</param>
+        /// <param name="till">specified height</param>
+        /// <returns>Whether deposit lock height was successfully updated.</returns>
+        [ContractMethod(CpuFee = 1 << 15, RequiredCallFlags = CallFlags.States)]
+        public bool LockDepositUntil(ApplicationEngine engine, UInt160 addr, uint till)
+        {
+            if (!engine.CheckWitnessInternal(addr)) return false;
```

Comment:
should be different to this.Hash, otherwise a notary node can lock deposit of the native contract

Reply:
> deposit of the native contract

Normally, notary contract doesn't have deposit for itself. Technically, it's not prohibited by `onNEP17Payment` rules, but it will cost some real GAS for the depositer, and this gas will be spent for nothing in fact. 

> otherwise a notary node can lock deposit 

For that Notary contract should be included into the list of transaction's signers with non-`None` witness scope which is prohibited by `verify` method of the contract:
https://github.com/neo-project/neo/blob/bb153eb993d963b1a19336a1bfd26ed877053b9d/src/Neo/SmartContract/Native/Notary.cs#L108

Reply:
Mmm not sure if it's the caller... but it should not be the caller..

Reply:
> a notary node can lock deposit of the native contract

To lock someone's deposit, the transaction should be signed by the deposit owner, it's a requirement, we have `CheckWitness` call here. But Notary contract sender/signer is forced to have `None` witness scope. So witness check for notary contract will fail here anyway respectively if it's a caller or not.

So I don't see a problem here.